python/ray/serve/tests/test_http_state.py

```diff
@@ -3,172 +3,517 @@
+@patch("ray.serve._private.http_state.PROXY_HEALTH_CHECK_PERIOD_S", 0.1)
+def test_http_proxy_state_update_healthy_check_health_always_timeout():
+    """Test calling update method on HTTPProxyState when the proxy state is HEALTHY and
+    when the ready call always timed out.
+
+    The proxy state started with STARTING. After update is called and ready call
+    succeeded, the status will change to HEALTHY. After the next few check_health calls
+    takes very long time to finished, the status will eventually change to UNHEALTHY
+    after all retries have exhausted.
+    """
+
+    @ray.remote(num_cpus=0)
+    class NewMockHTTPProxyActor:
+        async def ready(self):
+            return json.dumps(["mock_worker_id", "mock_log_file_path"])
+
+        async def check_health(self):
+            await asyncio.sleep(100)
 
-    ray.shutdown()
+    proxy_state = _create_http_proxy_state(proxy_actor_class=NewMockHTTPProxyActor)
+
+    # Continuously trigger update. The status should change from STARTING to HEALTHY
+    # when ready.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+    first_check_time = proxy_state._last_health_check_time
+
+    # Ensure the proxy status is HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
+
+    # Continuously trigger update and status continue to be UNHEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.UNHEALTHY,
+    )
+    # Ensure the proxy status is still UNHEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.UNHEALTHY
+
+    # Ensure the check time have changed since the last update
+    assert first_check_time != proxy_state._last_health_check_time
+
+    # Ensure _consecutive_health_check_failures is correct
+    assert proxy_state._consecutive_health_check_failures == 3
+
+
+@patch("ray.serve._private.http_state.PROXY_HEALTH_CHECK_PERIOD_S", 0.1)
+def test_http_proxy_state_update_unhealthy_check_health_succeed():
+    """Test calling update method on HTTPProxyState when the proxy state is UNHEALTHY
+    and when the check_health call succeeded.
+
+    The proxy state started with UNHEALTHY. After the next period of check_health
+    call, the status changes to HEALTHY.
+    """
+    proxy_state = _create_http_proxy_state()
+    proxy_state.set_status(status=HTTPProxyStatus.UNHEALTHY)
+
+    # Ensure the proxy status is UNHEALTHY.
+    assert proxy_state.status == HTTPProxyStatus.UNHEALTHY
+
+    # Continuously trigger update and status should change to HEALTHY.
+    wait_for_condition(
+        condition_predictor=_check_proxy,
+        state=proxy_state,
+        status=HTTPProxyStatus.HEALTHY,
+    )
+
+    # Ensure the proxy status is still HEALTHY after updates.
+    assert proxy_state.status == HTTPProxyStatus.HEALTHY
```

Comment:
Same here, I think this is redundant since you haven't called `.update()` in between the `wait_for_condition` succeeding and this assert.

Reply:
Yep, I did this in quite a few other places. But let me know if you feel strongly